internal/pkg/crypto/randomness.go

```diff
@@ -13,22 +14,22 @@ import (
 
 // RandomnessSource provides randomness to actors.
 type RandomnessSource interface {
-	Randomness(tag crypto.DomainSeparationTag, epoch abi.ChainEpoch, entropy []byte) (abi.Randomness, error)
+	Randomness(ctx context.Context, tag crypto.DomainSeparationTag, epoch abi.ChainEpoch, entropy []byte) (abi.Randomness, error)
 }
 
 type RandomSeed []byte
 
 type ChainSampler interface {
-	Sample(epoch abi.ChainEpoch) (RandomSeed, error)
+	Sample(ctx context.Context, epoch abi.ChainEpoch) (RandomSeed, error)
 }
 
 // A randomness source that seeds computations with a sample drawn from a chain epoch.
 type ChainRandomnessSource struct {
 	Sampler ChainSampler
 }
 
-func (c *ChainRandomnessSource) Randomness(tag crypto.DomainSeparationTag, epoch abi.ChainEpoch, entropy []byte) (abi.Randomness, error) {
-	seed, err := c.Sampler.Sample(epoch)
+func (c *ChainRandomnessSource) Randomness(ctx context.Context, tag crypto.DomainSeparationTag, epoch abi.ChainEpoch, entropy []byte) (abi.Randomness, error) {
```

Comment:
nit: It's unclear why we separate "entropy blending" from "epoch blending".  Do we ever use the raw output of sampler?  It seems like blendEntropy should be a subfunction on the sampler and should do epoch mixing along with entropy bytes and tag mixing.

Reply:
This is how I understand it from @sternhenri . I agree with you from a software org point of view it seems weird to first hash the ticket, then hash (that + epoch), and then hash (that + tag + entropy).

Reply:
small thing: first hashing the ticket makes sense from a crypto standpoint because sign then hash is one conceptual operation.

Agree about the next two hashes being weird but I am totally ok with taking this requirement from the spec.

Even taking the two hashes from the spec from our software point of view it still makes more sense to me for one thing/place to be doing all of the levels of hashing.  I don't think there is anything that consumes partial outputs, so why leave multiple layers of indirection here?

Reply:
Thanks. We can fix the organization later if it proves possible.

> I don't think there is anything that consumes partial outputs

I haven't hooked up ticket generation yet and am pretty worried it's different.